shadowsheep1/io-app
# [3] Refresh the user profile's data from its API call when the new Profile screen is mounted

## Short description
This PR calls the user profile backend API [`/profile`](https://github.com/pagopa/io-backend/blob/07c7d6cd6c27c10bf733faa490d1c2cf69aae374/api_backend.yaml#L505) in order to have a fresh user data during the new screen (see PR #2) mount, like described in #3.

| New profile page |
| - |
| <img src="https://user-images.githubusercontent.com/16268789/203809161-24532a3a-8f59-4aa1-8e9c-587e47e46ee1.png" width="150" /> |

## List of changes proposed in this pull request
This PR:
- Start over the currently open PR #2 to reuse the same new screen.
- Makes some clean up in code and project structure:
    - Put all the code and test files of the new "feature" under `./ts/feature/newProfile` folder.  With this structure:
```bash
newProfile
    - __test__
        - ProfileMainScreen.test.tsx # <- This is a rename of NewProfileMainScreen.test.tsx.
    - saga
        - profile.ts # <- new saga specific for this new screen profile data update. 
    - screens
        - ProfileMainScreen.tsx # <- This is a rename of NewProfileMainScreen.tsx, The Screen object has been renamed accordingly.
    - store
        - actions
            - profile.ts # <- new redux actions specific for this new screen profile data update.  
```

When the **new profile screen** is _loaded for the first time_ it takes the profile data from the current redux store but at the same time, in background it tries to get a fresh user profile data from the backend server:

- If the API call succeed, then the UI is rehydrated with the fresh data.
- If the API call fails for some reason the user is notified and a retry button is presented.
- If the API calls report a `401` the user is de-authenticated.

| API call loading state | API Call error state |
| - | - |
|  <img src="https://user-images.githubusercontent.com/16268789/203808793-02bfc43e-2a57-41be-9e70-84644c40e4d5.png" width="150" />  |  <img src="https://user-images.githubusercontent.com/16268789/203808871-9c13dd08-e2c6-4e18-a36e-a2b0baa26cee.png" width="150" />  |

## How to test
- Run the app and then navigate to the new profile page by tapping the profile icon on the bottom bar. What you see is the new profile pages. If you are pointing to the [`io-api-dev-server`](https://github.com/pagopa/io-dev-api-server) you should see a single call `GET /api/v1/profile` with status code  `200` or  `304` on console, or use Flipper/Reactotron. If you are pointing to the production server use **Flipper** or **Reactotron** to see this single api call to the server.
